File: LeetCode/Medium/1609-even-odd-tree/1609-even-odd-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isEvenOddTree(TreeNode* root) {
        if (root == nullptr || root->val % 2 == 0) {
            return false;
        }

        std::queue<TreeNode*> q;
        q.push(root);
        bool isOdd = true;

        while (!q.empty()) {
            int size = q.size();
            int prev = isOdd ? INT_MIN : INT_MAX;

            for (int i = 0; i < size; ++i) {
                TreeNode* curr = q.front();
                q.pop();

                if ((isOdd && (curr->val % 2 == 0 || curr->val <= prev)) ||
                    (!isOdd && (curr->val % 2 != 0 || curr->val >= prev))) {
                    return false;
                }

                if (curr->left != nullptr) {
                    q.push(curr->left);
                }
                if (curr->right != nullptr) {
                    q.push(curr->right);
                }

                prev = curr->val;
            }

            isOdd = !isOdd;
        }

        return true;
    }
};

File: LeetCode/Medium/1609-even-odd-tree/1609-even-odd-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isEvenOddTree(root *TreeNode) bool {
    if root == nil || root.Val%2 == 0 {
		return false
	}

	queue := []*TreeNode{root}
	isOdd := true

	for len(queue) > 0 {
		size := len(queue)
		prev := 0
		if isOdd {
			prev = -1 << 31
		} else {
			prev = 1<<31 - 1
		}

		for i := 0; i < size; i++ {
			curr := queue[i]

			if (isOdd && (curr.Val%2 == 0 || curr.Val <= prev)) ||
				(!isOdd && (curr.Val%2 != 0 || curr.Val >= prev)) {
				return false
			}

			if curr.Left != nil {
				queue = append(queue, curr.Left)
			}
			if curr.Right != nil {
				queue = append(queue, curr.Right)
			}

			prev = curr.Val
		}

		queue = queue[size:]
		isOdd = !isOdd
	}

	return true
}